extensions/mysql/lib/mysql.go

```diff
@@ -0,0 +1,439 @@
+		if err != nil {
+			tracelog.ErrorLogger.Fatalf("%+v\n", err)
+		}
+		return logFileName
+	}
+	tracelog.ErrorLogger.Fatalf("Failed to obtain current binlog file")
+	return ""
+}
+
+func getMySQLConnection() (*sql.DB, error) {
+	datasourceName := internal.GetSettingValue("WALG_MYSQL_DATASOURCE_NAME")
+	if datasourceName == "" {
+		datasourceName = "root:password@/mysql"
+	}
+	caFile := internal.GetSettingValue("WALG_MYSQL_SSL_CA")
+	if caFile != "" {
+		rootCertPool := x509.NewCertPool()
+		pem, err := ioutil.ReadFile(caFile)
+		if err != nil {
+			return nil, err
+		}
+		if ok := rootCertPool.AppendCertsFromPEM(pem); !ok {
+			return nil, fmt.Errorf("Failed to load certificate from %s", caFile)
+		}
+		mysql.RegisterTLSConfig("custom", &tls.Config{
+			RootCAs: rootCertPool,
+		})
+		if strings.Contains(datasourceName, "?tls=") || strings.Contains(datasourceName, "&tls=") {
+			return nil, fmt.Errorf("MySQL datasource string contains tls option. It can't be used with WALG_MYSQL_SSL_CA option")
+		}
+		if strings.Contains(datasourceName, "?") {
+			datasourceName += "&tls=custom"
+		} else {
+			datasourceName += "?tls=custom"
+		}
+	}
+	db, err := sql.Open("mysql", datasourceName)
+	return db, err
+}
+
+// TODO : unit tests
+// UploadFile compresses a file and uploads it.
+func (uploader *Uploader) UploadStream(fileName string, db *sql.DB, stream io.Reader) error {
+
+	binlogStart := getMySQLCurrentBinlogFile(db)
+	tracelog.DebugLogger.Println("Binlog start file", binlogStart)
+	timeStart := time.Now()
+
+	compressor := uploader.Compressor
+	pipeWriter := &internal.CompressingPipeWriter{
+		Input:                stream,
+		NewCompressingWriter: compressor.NewWriter,
+	}
+
+	pipeWriter.Compress(&internal.OpenPGPCrypter{})
+
+	backup := Backup{internal.NewBackup(uploader.UploadingFolder, fileName)}
+
+	dstPath := getStreamName(&backup, compressor.FileExtension())
+	tracelog.DebugLogger.Println("Upload path", dstPath)
+	reader := pipeWriter.Output
+
+	err := uploader.Upload(dstPath, reader)
+	tracelog.InfoLogger.Println("FILE PATH:", dstPath)
+	binlogEnd := getMySQLCurrentBinlogFile(db)
+	tracelog.DebugLogger.Println("Binlog end file", binlogEnd)
+
+	uploadStreamSentinel(&StreamSentinelDto{BinLogStart: binlogStart, BinLogEnd: binlogEnd, StartLocalTime: timeStart}, uploader, fileName+internal.SentinelSuffix)
+
+	return err
+}
+
+type StreamSentinelDto struct {
+	BinLogStart    string `json:"BinLogStart,omitempty"`
+	BinLogEnd      string `json:"BinLogEnd,omitempty"`
+	StartLocalTime time.Time
+}
+
+func uploadStreamSentinel(sentinelDto *StreamSentinelDto, uploader *Uploader, name string) error {
+	dtoBody, err := json.Marshal(*sentinelDto)
+	if err != nil {
+		return err
+	}
+
+	uploadingErr := uploader.Upload(name, bytes.NewReader(dtoBody))
+	if uploadingErr != nil {
+		tracelog.ErrorLogger.Printf("upload: could not upload '%s'\n", name)
+		tracelog.ErrorLogger.Fatalf("StorageTarBall finish: json failed to upload")
```

Comment:
You have to either fatal or return here, but not both. I strongly recommend you using return.